cratetorrent/src/disk/io/file.rs

```diff
@@ -1,19 +1,62 @@
 use std::{
     fs::{File, OpenOptions},
-    os::unix::io::AsRawFd,
     path::Path,
+    io::ErrorKind
 };
 
-use nix::sys::uio::{preadv, pwritev};
-
 use crate::{
     disk::error::*,
-    iovecs,
-    iovecs::{IoVec, IoVecs},
     storage_info::FileSlice,
     FileInfo,
 };
 
+use cfg_if::cfg_if;
+
+#[cfg(target_os = "linux")]
+use crate::iovecs::{self, IoVec, IoVecs};
+
+
+// a convenience macro for handling ErrorKind::Interrupted,
+// zero-byte reads and the like
+macro_rules! unwrap_read_res {
+    ($res:expr) => {
+        match $res {
+            Err(e) if e.kind() == ErrorKind::UnexpectedEof => return Err(ReadError::MissingData),
+            Err(e) if e.kind() == ErrorKind::Interrupted => continue,
+            Err(e) => return Err(e.into()),
+
+            // If there was nothing to read from file, it means we tried to
+            // read a piece from a portion of a file not yet downloaded or
+            // otherwise missing
+            Ok(0) => return Err(ReadError::MissingData),
+
+            Ok(written) => written,
+        }
+    }
+}
+
+// a convenience macro for handling ErrorKind::Interrupted,
+// zero-byte writes and the like
+macro_rules! unwrap_write_res {
```

Comment:
nitpick: I wonder if this and the above could just live in `From` impls for the read and write error types, since they're distinct anyway?

Reply:
`Interrupted => continue`, so no, not without making it more inconvenient to use.